documentation/specs/src/economics/fee-system.md

```diff
@@ -1,7 +1,143 @@
+
+Since the whitelist can be changed via governance, transactions could fail these checks in the block where the whitelist change happens. For `mempool_check`, the checks could reject transactions that may become valid in the future or vice-versa: since we can assume a slow rate of change for these parameters and mempool and block space optimizations are a priority, it is up to the clients to track any changes of these parameters and act accordingly.
+
+## Gas accounting
+
+We provide a mapping between all the whitelisted transactions and VPs to their cost in gas units. Being the cost hardcoded, it is guaranteed that the same transaction will always require the same amount of gas: since the price per gas is controlled via governance, though, the price for the same transaction may vary in time. A transaction is also charged with the gas required by the validity predicates that it triggers.
+
+Gas accounting is about preventing a transaction from exceeding two gas limits:
+
+1. Its own `GasLimit` (declared in the wrapper transaction)
+2. The gas limit of the entire block
+
+### Wrapper GasLimit
+
+The protocol injects a gas counter in each transaction and VP to be executed which allows monitoring of the exact amount of gas utilized. To do so, the gas meter simply checks the hash of the transaction or VP against the table in storage to determine which one it is and, from there, derives the amount of gas required.
+
+To perform the check we need the limit which was declared by the corresponding wrapper transaction: this limit should be saved in storage together with the queue of encrypted transactions for easy access.
+
+Since the hash can be retrieved as soon as the transaction gets decrypted, we can immediately check whether the `GasLimit` set in the corresponding wrapper is enough to cover this amount. This check, though, is weak because we also need to keep in account the gas required for the involved VPs which is hard to determine ahead of time: this is just an optimization to short-circuit the execution of transactions whose gas limit is not enough to cover for even the tx itself.
+
+When executing the VPs the procedure is the same and, again, since we know ahead of time the gas required by each VP we can immediately terminate the execution if it overshoots the limit.
+
+In any case, if the gas limit is exceeded, the transaction is considered invalid and all the modifications applied to the WAL get discarded. This doesn't affect the other transactions which can be executed normally (see the following section).
```

Comment:
I would just explicitly add here that the gas used for the tx that exceed its gas limit is still added to the used gas in the block

Reply:
Actually, since the block gas limit has to be compared against the the `GasLimit` declared in the `WrapperTx` we don't need to monitor the actual gas usage of an `InnerTx` against the block gas limit but only against the declared gas limit of the corresponding wrapper. So the gas meter could just check this constraint and avoid checking for the block gas limit since that is indirectly checked by the combination of these:
- checking in `process_proposal` that the cumulated `GasLimit` of alle the wrappers does not exceed the block gas limit
- preventing every inner from exceeding the gas limit allocated by its wrapper